File: firebase/functions/src/entity/BookLibrary.ts
import { Entity, Column, ManyToOne, BaseEntity, PrimaryGeneratedColumn } from "typeorm";
import { Book } from "./Book";
import { Library } from "./Library";

@Entity()
export class BookLibrary extends BaseEntity {
    @PrimaryGeneratedColumn()
        id:number;
    
    @Column({
        nullable: true}) 
    user_note: string;
    
    @Column() 
    private_book: boolean;

    @Column() 
    loanable: boolean;

    @Column() 
    reading: boolean;

    @Column() 
    loaned: boolean;

    @Column() 
    unpacked: boolean;

    @Column({
        nullable: true}) 
    rating: number;

    @ManyToOne (type => Library, library => library.bookLibrary, {onDelete: "CASCADE"})
    library: Library;

    @ManyToOne (type => Book, book => book.bookLibrary, {onDelete: "CASCADE"})
    book: Book;

}

File: lib/models/bookLibrary.dart
import '../models/book.dart';

class BookLibrary {
  int id;
  String notes;
  bool private;
  bool loanable;
  bool currentlyreading;
  bool checkedout;
  bool unpacked;
  int rating;
  Book book;

  BookLibrary(
      {this.book,
      this.notes,
      this.private,
      this.loanable,
      this.rating,
      this.currentlyreading,
      this.checkedout,
      this.unpacked,
      this.id});

  Map<String, dynamic> toJson() => {
        'user_note': notes,
        'private_book': private,
        'id': id,
        'loanable': loanable,
        'rating': rating,
        'reading': currentlyreading,
        'loaned': checkedout,
        'unpacked': unpacked
      };

  BookLibrary.fromJson(Map<dynamic, dynamic> record) {
    notes = record['user_note'];
    private = record['private_book'];
    loanable = record['loanable'];
    rating = record['rating'];
    currentlyreading = record['reading'];
    checkedout = record['loaned'];
    unpacked = record['unpacked'];
    id = record['id'];
    book = record['book'] != null ? Book.fromJson(record['book']) : null;
  }

  int get lbid {
    return id;
  }

  String get userNotes {
    return notes;
  }

  bool get secretBook {
    return private;
  }

  bool get loanOption {
    return loanable;
  }

  bool get curReading {
    return currentlyreading;
  }

  bool get isBorrowed {
    return checkedout;
  }

  bool get unpackedBook {
    return unpacked;
  }

  int get starRating {
    return rating;
  }
}
